torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -0,0 +1,36 @@
+#pragma once
+
+#include <c10/util/Exception.h>
+#include <torch/data/datasets/base.h>
+#include <torch/data/example.h>
+
+namespace torch {
+namespace data {
+namespace datasets {
+/// A dataset that supports loading an entire chunk of data.
+///
+/// A chunk could be an entire file, such as an audio data file or an image,
+/// or part of a file in the case of a large text file split based on seek
+/// positions. ChunkDataSet extends the DataSet functionality to read an
+/// entire chunk at once.
+template <
+    typename Self,
+    typename Batch = std::vector<Example<>>,
+    typename BatchRequest = ArrayRef<size_t>>
+class ChunkDataSet : public BatchDataset<Self, Batch, BatchRequest> {
```

Comment:
I don't think this class is necessary. The point of my PR to generalize the index type, https://github.com/pytorch/pytorch/pull/12960, was to allow you to create chunk datasets since you can just subclass `BatchDataset` and set the `BatchRequest` type to whatever you need. That's what confuses me about this whole `ChunkDataset`. Basically, you're suggesting `MyChunkDataset` would look like this:

```cpp
class MyChunkDataSet : public torch::data::datasets::ChunkDataSet<MyChunkDataSet> {
 public:
  Batch read_chunk(size_t chunk_index) override { /* fetch batch at chunk_index */ }
};
```

but we could also just have written

```cpp
class MyChunkDataSet : 
  public torch::data::datasets::BatchDataset<MyChunkDataSet, std::vector<Example>, size_t /* !!!! */> {
 public:
  Batch get_batch(size_t chunk_index) override { /* fetch batch at chunk_index */ }
};
```

and you'd get the exact same API, but now we can delete this whole class?

It's not that I don't want to support `ChunkDataset`. It's that I implemented the generalization of the index type precisely because you let me know that you want arbitrary index types. So now I don't see the value of `ChunkDataset` anymore. For example, what is the difference between the custom `get_batch()` in the second class above, and `read_chunk()` in the first example. None really, practically speaking, right?

Reply:
I think the confusion is mainly because you are assuming chunk==batch. Typically, a chunk is a much larger amount of data or Examples. Say 1000 Examples and a batch is set to 100. We want to pre-load an entire chunk when we read data from the underneath storage, but then divide it to 10 batches for training. 

So now, as you have mentioned above, we can use the BatchDataSet to load the chunk, but we cannot return the whole chunk to the user or apply collate() or any other map() operation on the whole chunk. We still need to do those at the individual small batch level. This means we need two APIs, one to load the whole chunk and one to load the batch. 

The main reason for having two APIs is to support the map() operation. To use map()/collate() on a dataset, the data must come out from the get_batch() API, which is run on the dataloader threads. We want preloading of chunks(bigger blocks) and let user only worry about that, and hence the second API. Hope this clarifies.

Reply:
I still don’t understand why the previous interface is insufficient. Preloading more data and caching it in the dataset object is something that’s very easy to do using the existing API, so I don’t think this justifies the added complexity of this solution.

Reply:
Could we discuss this over a call. That would be more efficient.

Reply:
@apaszke This is exactly what we try to achieve and hence the minimum change to the dataloader. The two options we discussed was whether to change the dataloader or dataset and decided to move the preloading to the dataset and keep the dataloader as close to what it is today. To get this with the existing APIs, we need the changes in the PR. For example, Dataset has a single key API, get_batch() we will implement all the preloading under this API, but then we would like user's to implement the data parsing in read_chunk() API. We don't want the dataset to be copied to threads or someone specify a size to a sampler because those should come after reading data. This PR is will get us the flexibility to do those using the existing dataloader. Please let me know if you have specific questions.

Reply:
We had an offline discussion with Peter and discussed our proposal in detail. We ended up with two main tasks, (I) for the dataloader constructor overload, Peter is trying to see if we can move the dataset without copying via a wrapper.  (ii) Peter also pointed us that the current BatchSizeSampler will not stop the dataloder even after the dataset is exhausted, so we are looking into a fix for that. Once we have those, we will provide an update to the PR.